### PR TITLE
fix: correctly resolve duplicate struct names in json abi

### PR DESCRIPTION
**Pull request by klkvr**
## Motivation

Closes https://github.com/alloy-rs/core/issues/660

## Solution

I've added second abi preprocessing method `JsonAbi::rename_internal_structs`, which goes over ABI, resolves structs that have a same name but different components sets, and renames them into `<struct name>0`, `<struct name>1`, etc

It should be also possible to key them by `contract` instead, however, there are some edge cases related to aliases with such approach. E.g. `import {Struct as Struct1} from "..."; import {Struct as Struct2} from "...";` will result in two different structs referenced as `Struct` in ABI. Same is possible with `Lib.Struct` if different `Lib`s are aliased.

This edge case is kind of weird, so I am open to changing impl if we consider performance improvements worth it.

Also not sure if we want this to happen on the level of `JsonAbi::to_sol` instead of only for `sol!` used code

## PR Checklist

- [x] Added Tests
- [x] Added Documentation
- [ ] Breaking changes

### CODE REVIEW

File: crates/json-abi/src/abi.rs [hunk 1]
```diff
@@ -608,3 +649,82 @@ impl<'de> Visitor<'de> for ContractObjectVisitor {
         Ok(ContractObject { abi, bytecode, deployed_bytecode })
     }
 }
+
+/// A [crate::Visitor] implementation that collects a mapping from struct name to set of components
+/// various structs with such name have.
+#[derive(Debug, Default, Clone)]
+struct Collector<'a> {
+    unique_struct_components: BTreeMap<&'a str, BTreeSet<Vec<&'a str>>>,
```
Comment (mattsse) [inline crates/json-abi/src/abi.rs:657]: perhaps worth using smallvec here for the `Vec` because we only expect a few projections?

File: crates/json-abi/src/abi.rs [hunk 2]
```diff
@@ -246,6 +253,40 @@ impl JsonAbi {
...
+            .map(|(name, components)| {
+                (
+                    name.to_string(),
+                    components
+                        .into_iter()
+                        .enumerate()
+                        .map(move |(idx, components)| {
+                            let components =
+                                components.into_iter().map(ToString::to_string).collect::<Vec<_>>();
+                            (components, format!("{name}{idx}"))
```
Comment (mattsse) [inline crates/json-abi/src/abi.rs:279]: should we start at 0 here or at 1?

File: crates/json-abi/src/abi.rs [hunk 3]
```diff
@@ -608,3 +649,82 @@ impl<'de> Visitor<'de> for ContractObjectVisitor {
         Ok(ContractObject { abi, bytecode, deployed_bytecode })
     }
 }
+
+/// A [crate::Visitor] implementation that collects a mapping from struct name to set of components
+/// various structs with such name have.
```
Comment (mattsse) [inline crates/json-abi/src/abi.rs:654]: ah okay, I understand how this works, first we traverse the abi and collect unique names and their paths

**Review by mattsse — CHANGES_REQUESTED** [review]
cool, I believe this works and is easy to follow.

can we add another test case for https://github.com/foundry-rs/foundry/issues/8444

imo we can ignore this edge case `import {Struct as Struct2}`

ptal @DaniPopes

File: crates/json-abi/src/abi.rs [hunk 4]
```diff
@@ -246,6 +253,40 @@ impl JsonAbi {
         }
     }
 
+    /// Goes over all structs referenced as internal types in the ABI and changes them to global
+    /// structs.
+    ///
+    /// Namespace for global struct names is managed and in cases of duplicates structs are renamed.
+    pub fn rename_internal_structs(&mut self) {
```
Comment (DaniPopes) [inline crates/json-abi/src/abi.rs:260]: I don't think any of this should be exposed publicly, it should only be a sol! implentation detail

**Review by DaniPopes — COMMENTED** [review]
I don't really like renaming because this doesn't tackle the underlying problem that namespaces are not handled in sol!, e.g. this only fixes json input

I have a branch open where I make namespaced output work, but got stuck on weird compile errors https://github.com/alloy-rs/core/tree/dani/to-sol-contract-specifiers

File: crates/json-abi/src/abi.rs (hunk 2 above)
Comment (klkvr) [reply crates/json-abi/src/abi.rs:279]: function overloads start with 0, so I've just made it similar

just noticed that functions are named as `function_0`, `function_1`, should we include underscore here as well?

**Comment by klkvr** [comment]
> I have a branch open where I make namespaced output work, but got stuck on weird compile errors https://github.com/alloy-rs/core/tree/dani/to-sol-contract-specifiers

yeah, such approach indeed is more flexible

current branch compiles fine for me, do you have more fixes for sol! as wip?

from diff and a quick test it seems that updated `Type::Custom` branch in `rec_expand_type` does most of the job on the `sol!` side? would be happy to finish this if you'll point out what's left/not working yet

**Comment by DaniPopes** [comment]
It fails to compile in tests in crates/sol-types/tests/macros/sol/json.rs, but expanding the output and compiling it separately is fine

It only handles the ToSol side, expanding libraries, but namespaces are not handled in the sol! internal representation, so function signatures are wrong

Probably best to just continue that branch

**Comment by klkvr** [comment]
> It fails to compile in tests in crates/sol-types/tests/macros/sol/json.rs, but expanding the output and compiling it separately is fine

fyi moving `sol!` invocation out of the test function scope fixes compilation, will look into why it's happening

**Comment by klkvr** [comment]
one of the issues seems to be a limitation of function-level modules

e.g. this code doesn't work
```rust
fn test_fn() {
    pub mod inner1 {
        pub struct InnerStruct {}
    }
    
    pub mod inner2 {
        use super::*;
        fn some_fn() {
            inner1::InnerStruct {};
        }
    }
}
```

and it seems that there's no way to reach other modules defined inside of a function from inside of another module

**Comment by DaniPopes** [comment]
Superseded by https://github.com/alloy-rs/core/pull/693, #694
